### PR TITLE
Fibonacci function, pt.2: basic calls

### LINKED ISSUE #3: Milestone: Fibonacci function
Recursive Fibonacci is a nice next milestone because it needs some control flow operators and calls, which should really help establish the shape of the backend.

The steps are:
 - Implement `if`+`else`+`end`.
     - We'll need a stack for tracking nested control flow constructs. To start with there's only one kind of control-flow stack entry, for if/else/end, but eventually we'll have more. It can hold [`DynamicLable`](https://docs.rs/dynasmrt/0.2.3/dynasmrt/struct.DynamicLabel.html)s to keep track of the labels for branching.
     - The result value of an `if` can be carried by the stack just like normal operator results. (For now; later with on-the-fly register allocation it can be more sophisticated.)
 - Implement calls
    - This is a good time to generalize the handling of function signatures. I think we can just collect all the `FuncType`s from the type section into a `Vec` and pass that around for now.
    - Optionally, to keep things simple for this step, we could limit support to just 6 integer arguments, to avoid having to deal with stack arguments just yet.

I've left out a lot of the low-level details here; feel free to ask for more detail!

### PR DESCRIPTION
**Pull request by pepyakin**
Second part of and, hopefully, closes #3 

For that, I started reading the type section and function section, to aggregate the information about function types. I decided to introduce a `TranslationContext` which holds all this data and, may be, will answer queries about global types, and other module specific information.

Then, I had to update the `CodeGenSession` to keep `DynamicLabel`s of functions. I did this in order to simplify the forward calling of functions. However, there seems to be no way of getting the `AssemblyOffset` from `DynamicLabel` so I had to keep them as well.

Along the way, I have implemented `i32.const` because... well, it turned out that it is very useful instruction : )

And of course, I implemented the `call` instruction. For now, I decided to implement it via two backend functions:

- `pass_outgoing_args`, which takes a function's `arity`.
- `call_direct`, which takes `function_index` and `return_arity`

I thought that this separation may be useful, because sometimes we find ourselves in a situation when we already have arguments in their places or we need some custom handling.

so basically, backend now knows and uses function starts.

### CODE REVIEW

File: src/backend.rs [hunk 1]
```diff
@@ -264,6 +270,14 @@ pub fn store_i32(ctx: &mut Context, local_idx: u32) {
     ctx.regs.release_scratch_gpr(gpr);
 }
 
+pub fn literal_i32(ctx: &mut Context, imm: i32) {
+    let gpr = ctx.regs.take_scratch_gpr();
+    dynasm!(ctx.asm
+        ; mov Rd(gpr), imm
```
Comment (pepyakin) [inline src/backend.rs:298]: I'm not entirely sure if this is correct...

**Comment by eira-fransham** [comment]
For calls, is it acceptable to do a post-processing pass (as `dynasm` appears to do for dynamic labels)? It would still make runtime linear, although it would no longer be single-pass. I think for us that's acceptable, although I'd have to confirm that.

File: src/backend.rs [hunk 2]
```diff
@@ -346,7 +392,12 @@ pub fn prologue(ctx: &mut Context, stack_slots: u32) {
 }
 
 pub fn epilogue(ctx: &mut Context) {
-    assert_eq!(ctx.sp_depth, StackDepth(0), "imbalanced pushes and pops detected");
+    // TODO: This doesn't work with stack alignment.
+    // assert_eq!(
+    //     ctx.sp_depth,
+    //     StackDepth(0),
+    //     "imbalanced pushes and pops detected"
+    // );
```
Comment (pepyakin) [inline src/backend.rs:422]: I'm not sure how to cleanly fix this one...

**Comment by pepyakin** [comment]
I just realized that code in this PR doesn't align stack to the 16-byte boundary before a call. Given that we already have code based on this PR and that this isn't strict requirement I'd propose to make a follow-up on that.

File: src/backend.rs [hunk 3]
```diff
@@ -117,33 +109,54 @@ fn abi_loc_for_arg(pos: u32) -> ArgLocation {
 
 pub struct CodeGenSession {
     assembler: Assembler,
-    func_starts: Vec<AssemblyOffset>,
+    func_starts: Vec<(Option<AssemblyOffset>, DynamicLabel)>,
```
Comment (sunfishcode) [inline src/backend.rs]: FWIW, I discussed this with @CensoredUsername on irc a while ago, and they said that in the future there may be a way to get dynasm to tell us the AssemblyOffset for a resolved DynamicLabel, but that for now we should ago ahead and duplicate the info like so.

File: src/backend.rs [hunk 4]
```diff
@@ -117,33 +109,54 @@ fn abi_loc_for_arg(pos: u32) -> ArgLocation {
...
     }
 
-    pub fn new_context(&mut self) -> Context {
-        let start_offset = self.assembler.offset();
-        self.func_starts.push(start_offset);
+    pub fn new_context(&mut self, func_idx: u32) -> Context {
+        {
+            let func_start = &mut self.func_starts[func_idx as usize];
+
+            // At this point we now the exact start address of this function. Save it
```
Comment (sunfishcode) [inline src/backend.rs]: Missing word "know"?

File: src/backend.rs (hunk 1 above)
Comment (sunfishcode) [reply src/backend.rs:298]: Looks ok to me. Eventually we'll want to optimize constants in various cases, but we can figure that out later :-).

File: src/backend.rs [hunk 5]
```diff
@@ -327,8 +341,40 @@ pub fn copy_incoming_arg(ctx: &mut Context, arg_pos: u32) {
...
+
+pub fn call_direct(ctx: &mut Context, index: u32, return_arity: u32) {
+    assert!(return_arity == 0 || return_arity == 1);
+
+    let label = &ctx.func_starts[index as usize].1;
+    dynasm!(ctx.asm
+        ; call =>*label
     );
+
+    if return_arity == 1 {
```
Comment (sunfishcode) [inline src/backend.rs:394]: Could you add an `assert!(return_arity <= 1)` for now, so that we don't silently misbehave if someone starts working on multi-value support?

File: src/backend.rs (hunk 2 above)
Comment (sunfishcode) [reply src/backend.rs:422]: I'm ok dealing with this later, but eventually we will need to remember how much space we used for stack alignment or other things, which should let us reenable this assert.

**Review by sunfishcode — COMMENTED** [review]
Horray, Fibonacci :-)

For ensuring stack alignment at calls, let's add a TODO comment for now.

We may be getting close to the point where we should start looking at wasmtime-environ and replacing `TranslationContext` with [wasmtime-environ's Module](https://github.com/CraneStation/wasmtime/blob/master/lib/environ/src/module.rs#L132). That's not available on crates.io yet, but if you think that makes sense, we can start that up.

File: src/backend.rs (hunk 3 above)
Comment (pepyakin) [reply src/backend.rs]: Extracted as a named struct and commented it.

File: src/backend.rs (hunk 5 above)
Comment (pepyakin) [reply src/backend.rs:394]: do you mean another one? : )

**Comment by pepyakin** [comment]
> We may be getting close to the point where we should start looking at wasmtime-environ and replacing TranslationContext with wasmtime-environ's Module. 

Hm, I can't say that I fully understand how can we do that.

File: src/backend.rs (hunk 5 above)
Comment (sunfishcode) [reply src/backend.rs:394]: Oops, you're right :-).

**Review by sunfishcode — COMMENTED** [review]
Alright, things are moving forward here!

**Comment by sunfishcode** [comment]
@pepyakin I've now published a wasmtime-environ crate on crates.io which should make it easier to start using wasmtime-environ to handle the module declaration sections. Here's how I think it can look:

```rust
        let mut module = Module::new();
        let tunables = Tunables::default();
        let environ = ModuleEnvironment::new(isa, &mut module, tunables);

        let translation = environ
            .translate(&data)
            .map_err(|error| ActionError::Compile(CompileError::Wasm(error)))?;
```

where &data is the incoming `&[u8]`. (This is adapted from the code in lib/execute/src/instance_plus.rs, in InstancePlus::new, if you want more context.)

Then, `translation.lazy.function_body_inputs` contins the `&[u8]`s of the function bodies, and `translation.module` contains the Module information, including function signatures and so on.

There's room for these APIs to be improved, but I think this is a good direction to start heading, as it'll take us closer toward being able to integrate with the runtime pieces of wasmtime, which we'll need to run real programs with lightbeam. What do you think?

**Comment by eira-fransham** [comment]
For `fn translate`, can we take any type `R: Read` instead of just `&[u8]`? It's not useful for Cranelift but for Lightbeam it could lead to way better performance when reading from a file because you don't need to load the whole file into memory first.

**Comment by sunfishcode** [comment]
Yes, I agree we want to move away from having a single `&[u8]` for the whole module.
